src/Microsoft.DotNet.Cli.Utils/FrameworkDependencyFile.cs

```diff
@@ -45,6 +46,36 @@ public string GetNetStandardLibraryVersion()
                 ?.Version;
         }
 
+        public bool TryGetMostFitRuntimeIdentifier(IEnumerable<string> candidateRuntimeIdentifiers, out string mostFitRuntimeIdentifier)
+        {
+            mostFitRuntimeIdentifier = null;
+
+            if (!SupportsCurrentRuntime())
+            {
+                return false;
+            }
+
+            RuntimeFallbacks runtimeFallbacks
+                = DependencyContext.RuntimeGraph.Where(g => g.Runtime == RuntimeEnvironment.GetRuntimeIdentifier()).Single();
+
+            if (candidateRuntimeIdentifiers.Contains(runtimeFallbacks.Runtime))
+            {
+                mostFitRuntimeIdentifier = runtimeFallbacks.Runtime;
+                return true;
+            }
+
+            foreach (var r in runtimeFallbacks.Fallbacks)
+            {
+                if (candidateRuntimeIdentifiers.Contains(r))
```

Comment:
I know I said perf wasn't a main concern, and N is small in practice, but I'd still avoid Enumerable.Contains in a loop with arbitrary iterations. I'd new up a HashSet up-front. It's just one more line of code.

I'm guessing the HashSet allocation won't be materially than the N enumerator allocations from Enumerable.Contains. I can back off if you measure it being significantly worse for common values of N. I'd rather some small overhead for small N but no risk of a multiplier by N getting out of hand. Maybe we've stopped adding version-specific RIDs, but the lists I see now grow by 2 or 3 with every OS release. Having that act as a multiplier could add up over time.

Reply:
Hmm, I guess the number of candidates doesn't grow as fast as the number of fallbacks. People probably won't add more than 3 shims. Still like to avoid having to think about risks of O(N^2) or O(N*M) if it's simple to avoid.

Reply:
is there a way to convert to HashSet and run Contains with ignorecasing?

We can make it a new type and override equal. Do you think it worth it?

Reply:
There's a constructor that takes an IEnumerable of items and an IEqualityComparer. Pass the candidates as the items and StringComparer.OrdinalIgnoreCase as the comparer.

Reply:
how to take it out? the return need to preserve the casing. candidate.Contains(machineRuntime) can only tell you "machineRuntime" is in, but we still need to return the candidate with casing.

Reply:
You can use `Dictionary<string, string>` as your lookup and TryGetValue instead of Contains:

``` C#
var candidateMap = candidates.ToDictionary(x => x, StringComparer.OrdinalIgnoreCase);

foreach (var fallback in fallbacks)
    if (candidateMap.TryGetValue(fallback, out var match))
        return match;
```